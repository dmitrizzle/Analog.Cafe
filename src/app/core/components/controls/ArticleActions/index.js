import React from "react"

import {
  getHumanDatestamp,
  getISODatestamp,
  getLunarDatestamp
} from "../../../utils/messages-"
import ArticleActionsWrapper from "./components/ArticleActionsWrapper"
import Link from "../Link"
import Options from "./components/Options"
import TimeStamp from "../../pages/Article/components/TimeStamp"

export const dateFactory = unix => {
  return {
    unix: unix,
    iso: getISODatestamp(unix),
    human: getHumanDatestamp(unix),
    lunar: getLunarDatestamp(unix)
  }
}
const DatePublished = props => {
  const dateModified =
    props.thisArticleEditDate &&
    props.thisArticleEditDate !== props.thisArticlePostDate
      ? dateFactory(props.thisArticleEditDate)
      : null
  const datePublished = dateFactory(props.thisArticlePostDate)
  return (
    <TimeStamp>
      <time title={`Published on ${datePublished.human}.`}>
        {datePublished.lunar}
      </time>
      {dateModified && [
        <span key="divider"> ✏︎ </span>,
        <time key="tiemstamp" title={`Edited on on ${dateModified.human}.`}>
          {dateModified.lunar}
        </time>
      ]}
    </TimeStamp>
  )
}

export default props => {
  return (
    <ArticleActionsWrapper>
      {props.thisArticlePostDate && <DatePublished {...props} />}

      {props.user &&
        props.user.status === "ok" &&
        (props.user.info.role === "admin" ||
          props.article.submittedBy.id === props.user.info.id) &&
        props.article.edits !== undefined &&
        props.article.edits.length > 0 && (
          <div style={{ lineHeight: "1em" }}>
            <strong>Edit History</strong>
            <br />
            {props.article.edits.map(edit => {
              const unix = new Date(edit.date * 1000)

              return (
                <React.Fragment key={edit.date}>
                  <small style={{ opacity: 0.5 }}>
                    {dateFactory(edit.date).human +
                      " " +
                      unix.getHours() +
                      ":" +
                      unix.getMinutes()}{" "}
                    – <Link to={`/author/${edit.id}`}>{edit.name}</Link>
                  </small>
                  <br />
                </React.Fragment>
              )
            })}
          </div>
        )}

      {props.article &&
        props.article.status === "published" &&
        typeof props.article.scheduledOrder === "undefined" && (
          <Options {...props} />
        )}
    </ArticleActionsWrapper>
  )
}
