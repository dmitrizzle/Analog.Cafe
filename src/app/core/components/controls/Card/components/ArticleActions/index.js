import React from "react"

import { Button, LinkButton } from "../../../Button"
import { CardFlattened, CardCaption } from "../../styles"
import {
  FacebookLinkButton,
  InstagramLinkButton,
  TwitterLinkButton
} from "../../../Button/components/SocialButtons"
import { PicturePlaceholder } from "../../../../vignettes/Picture/components/PicturePlaceholder"
import { QuickSubscribe } from "../../../../../../user/components/forms/Subscribe"
import { ROUTE_URL_ARTICLES } from "../../../../../constants/routes-article"
import { TimeStamp } from "../../../../styles/ArticleStyles"
import { authorNameList } from "../../../../../utils/authorship"
import { datestamp, lunar, percise } from "../../../../../utils/datestamp"
import { froth } from "../../../../../utils/image-froth"
import Link from "../../../Link"

const nextArticlePreload = nextArticle => {
  return {
    title: nextArticle.title,
    subtitle: nextArticle.subtitle,
    authors: nextArticle.authors,
    slug: nextArticle.slug,
    poster: nextArticle.poster,
    tag: nextArticle.tag
  }
}

// return
const ActionsCard = props => {
  if (props.mode !== "follow")
    return (
      <div>
        <CardFlattened>
          <QuickSubscribe
            subscribeFormCallback={props.subscribeFormCallback}
            stateOverwrite={props.subscribeForm}
          />

          {props.shareButtons ? (
            <FacebookLinkButton
              to="https://facebook.com/analog8cafe"
              onClick={event => {
                props.shareOnFacebook(event)
                // async load Google Analytics module
                import("react-ga").then(ReactGA => {
                  ReactGA.event({
                    category: "Campaign",
                    action: "ActionsCard.share_facebook"
                  })
                })
              }}
            >
              Share&nbsp;
            </FacebookLinkButton>
          ) : null}
          {props.shareButtons ? (
            <TwitterLinkButton
              to="https://twitter.com/analog_cafe"
              onClick={event => {
                props.shareOnTwitter(event)
                // async load Google Analytics module
                import("react-ga").then(ReactGA => {
                  ReactGA.event({
                    category: "Campaign",
                    action: "ActionsCard.share_twitter"
                  })
                })
              }}
            >
              Tweet
            </TwitterLinkButton>
          ) : null}
          {!props.hideShareButtons && (
            <Button
              onClick={props.revealShareButtons}
              inverse={props.shareButtons}
            >
              Share{" "}
              <span
                style={{ transform: "rotate(90deg)", display: "inline-block" }}
              >
                ⎋
              </span>
            </Button>
          )}
        </CardFlattened>
        {props.nextArticle &&
          props.nextArticle.slug && (
            <CardFlattened>
              <figure>
                <Link
                  to={ROUTE_URL_ARTICLES + "/" + props.nextArticle.slug}
                  onClick={() => {
                    props.nextArticleHeading(
                      nextArticlePreload(props.nextArticle)
                    )

                    // async load Google Analytics module
                    import("react-ga").then(ReactGA => {
                      ReactGA.event({
                        category: "Navigation",
                        action: "ActionsCard.next_article_picture"
                      })
                    })
                  }}
                >
                  <PicturePlaceholder frothId={props.nextArticle.poster}>
                    <img
                      src={
                        froth({ src: props.nextArticle.poster, size: "s" }).src
                      }
                      alt={props.nextArticle.title}
                    />
                  </PicturePlaceholder>
                </Link>
                <figcaption>
                  <CardCaption>
                    <span
                      style={{
                        display: "block",
                        fontSize: "0.8em",
                        lineHeight: "1.5em"
                      }}
                    >
                      <span style={{ opacity: "0.5" }}>Up next:</span>{" "}
                      <q>
                        {props.nextArticle.title}
                        {props.nextArticle.subtitle
                          ? " (" + props.nextArticle.subtitle + ")"
                          : null}
                      </q>{" "}
                      – {props.nextArticle.tag.replace(/-/g, " ")} by{" "}
                      {authorNameList(props.nextArticle.authors)}.
                    </span>
                  </CardCaption>
                </figcaption>
              </figure>
              <LinkButton
                style={{ margin: 0 }}
                to={ROUTE_URL_ARTICLES + "/" + props.nextArticle.slug}
                onClick={() => {
                  props.nextArticleHeading(
                    nextArticlePreload(props.nextArticle)
                  )

                  // async load Google Analytics module
                  import("react-ga").then(ReactGA => {
                    ReactGA.event({
                      category: "Navigation",
                      action: "ActionsCard.next_article_button"
                    })
                  })
                }}
              >
                Continue Reading <span>➢</span>
              </LinkButton>
            </CardFlattened>
          )}
      </div>
    )
  else
    return (
      <CardFlattened>
        <TwitterLinkButton
          to="https://twitter.com/analog_cafe"
          onClick={() => {
            // async load Google Analytics module
            import("react-ga").then(ReactGA => {
              ReactGA.event({
                category: "Campaign",
                action: "ActionsCard.follow_twitter"
              })
            })
          }}
        >
          Follow on Twitter
        </TwitterLinkButton>

        <FacebookLinkButton
          to="https://facebook.com/analog8cafe"
          onClick={() => {
            // async load Google Analytics module
            import("react-ga").then(ReactGA => {
              ReactGA.event({
                category: "Campaign",
                action: "ActionsCard.follow_facebook"
              })
            })
          }}
        >
          Follow on Facebook
        </FacebookLinkButton>

        <InstagramLinkButton
          to="https://instagram.com/analog_cafe"
          onClick={() => {
            // async load Google Analytics module
            import("react-ga").then(ReactGA => {
              ReactGA.event({
                category: "Campaign",
                action: "ActionsCard.follow_instagram"
              })
            })
          }}
        >
          Follow on Instagram
        </InstagramLinkButton>
      </CardFlattened>
    )
}

const DatePublished = props => {
  if (props.thisArticlePostDate)
    return (
      <TimeStamp
        dateTime={percise(props.thisArticlePostDate)}
        itemprop="datePublished"
        title={"Published on " + datestamp(props.thisArticlePostDate) + "."}
      >
        {lunar(props.thisArticlePostDate)}
      </TimeStamp>
    )
  else if (props.thisArticlePostDate) return null
  else return null
}

export default props => {
  return (
    <div style={{ clear: "both" }}>
      <DatePublished {...props} />
      <ActionsCard {...props} />
    </div>
  )
}
