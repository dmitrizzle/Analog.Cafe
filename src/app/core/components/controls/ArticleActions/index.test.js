import { shallow } from "enzyme"
import React from "react"
import "jest-styled-components"

import ArticleActions, { dateFactory } from "./"

it("Render ArticleActions, matches snapshot", () => {
  const element = shallow(<ArticleActions />)
  expect(element).toMatchSnapshot()
})
it("Produces correct date format", () => {
  expect(dateFactory(1529853719)).toEqual({
    unix: 1529853719,
    iso: "2018-06-24T15:21:59.000Z",
    human: "June 24, 2018",
    lunar: "24â˜¾6"
  })
})
