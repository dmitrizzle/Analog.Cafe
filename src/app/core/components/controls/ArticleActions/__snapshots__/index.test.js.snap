// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ArticleActions, matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Unknown />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        undefined,
        undefined,
        undefined,
      ],
      "style": Object {
        "clear": "both",
        "textAlign": "center",
      },
    },
    "ref": null,
    "rendered": Array [
      undefined,
      undefined,
      undefined,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          undefined,
          undefined,
          undefined,
        ],
        "style": Object {
          "clear": "both",
          "textAlign": "center",
        },
      },
      "ref": null,
      "rendered": Array [
        undefined,
        undefined,
        undefined,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
